File: leetcode/1115 - Print FooBar Alternately.cpp
class FooBar {
private:
    int n;

    std::mutex mutex;
    std::condition_variable cv;

    bool foo_ready = false;
    bool bar_ready = false;

public:
    FooBar(int n) {
        this->n = n;
    }

    void foo(const function<void()>& printFoo) {
        
        for (int i = 0; i < n; ++i) {
            {
                std::lock_guard<std::mutex> guard(mutex);
                printFoo();
                foo_ready = true;
                bar_ready = false;
            }

            cv.notify_one();
        	
            {
                std::unique_lock<std::mutex> guard(mutex);
                cv.wait(guard, [this]() { return this->bar_ready; });
            }
        }
    }

    void bar(const function<void()>& printBar) {
        
        for (int i = 0; i < n; ++i) {
            {
                std::unique_lock<std::mutex> guard(mutex);
                cv.wait(guard, [this]() { return this->foo_ready; });

                printBar();

                foo_ready = false;
                bar_ready = true;
            }

            cv.notify_one();
        }
    }
};

File: leetcode/1115 - Print FooBar Alternately.py
class FooBar:
    def __init__(self, n):
        self.n = n
        self.sem_foo = threading.Semaphore(0)
        self.sem_bar = threading.Semaphore(0)


    def foo(self, printFoo):
        
        for i in range(self.n):
            
            # printFoo() outputs "foo". Do not change or remove this line.
            printFoo()

            self.sem_foo.release()   # increment foo
            self.sem_bar.acquire()   # wait when bar is non zero


    def bar(self, printBar: 'Callable[[], None]') -> None:
        
        for i in range(self.n):
            self.sem_foo.acquire()   # wait when it is not 0 and decrement

            # printBar() outputs "bar". Do not change or remove this line.
            printBar()

            self.sem_bar.release()   # increment bar